### ISSUE TITLE
EncFS performance over NFS: 78Mbs/ raw, 7.7Mb/s encrypted

### ISSUE THREAD

**Issue by andrewparks** [vgough/encfs#237]
Scenario: one of the biggest potential uses for EncFS is for NAS, where you don't want to have to trust the NAS's own security.

Encryption Setup: Other than EncFS, no other encryption is being used. No NFS/network protocol encryption, no encryption within the NAS. 

Problem: Raw copy speeds are 10x faster than encrypted copy (write) speeds. Encrypted read speeds are full speed, so this only applies to writing.

Tested by copying 1gb.bin files from Mac to NAS with and without an EncFS filesystem.

Note that the performance issue only applies to copying (writing) 1GB files TO the NAS. Copying files back from an EncFS filesystem on the NAS back to the Mac happen at the full speed of 78MB/s.

System: Macbook Pro Retina Late 2013 2.8 GHz (i7-4558U) 
EncFS: 1.9.0
NAS: QNAP TS-453A w/ Intel N3160 1.6Ghz w/AES-NI. No I/O contention from any other users.
Network: Wired 1Gbit/s LAN via a single network switch
Network Protocol: NFS, mount params:
sudo mount -o rw,bg,hard,resvport,intr,nfc,tcp,async,rsize=32768,wsize=32768 -t nfs qnap:/raid1vol1 /nas/qnap-nfs

Any advice on improving performance? I've been experimenting with a wide number of NFS options and different network protocols (AFS/NFS/SMB) and 7.7Mb/s is the fastest I've ever been able to make it go. Thanks.

**Comment by rfjakob** [comment]
I cannot reproduce this on Linux 4.7.2. I get 53MB/s on the plain NAS and 43MB/s through EncFS.

Comparing the mount options, I noted that I have `rsize=131072,wsize=131072`, which is 4x your values.

**Comment by andrewparks** [comment]
Thanks so much for your reply. I did more testing, and found big differences in performance depending on rsize/wsize:

rsize 131072, wsize 131072: write @ 6.9mb/s
rsize 131072, wsize 131072: write @ 6.9mb/s
rsize 131072, wsize 131072: write @ 6.9mb/s
rsize 131072, wsize 131072: write @ 6.9mb/s
rsize 131072, wsize 131072: write @ 6.9mb/s

**Comment by andrewparks** [comment]
Thanks so much for your response, I tested and got the following results:

rsize 131072, wsize 131072: write @ 6.9 mb/s
rsize 4096, wsize 4096: write @ 39.4 mb/s
rsize 1024, wsize 1024: write @ 17.1 mb/s
rsize 8192, wsize 8192: write @ 13.3 mb/s
rsize 2048, wsize 2048: write @ 26.9 mb/s
rsize 3072, wsize 3072: write @ 26.3 mb/s
rsize unspecified, wsize unspecified: write @ 19.0 mb/s
rsize 4096, wsize 4096: (2nd test) write @ 40.3 mb/s, read @ 45.9 mb/s
rsize 131072, wsize 4096: write @ 9.26 mb/s, read @ 96.6 mb/s
rsize 8192, wsize 4096: write @ 30.7 mb/s, read @ 64.0 mb/s

Therefore the best tradeoff seems to be rsize 8192 / wsize 4096

Hope this helps anyone else with performance issues on OSX.

**Comment by andrewparks** [comment]
FYI, I've installed Fedora 24 (Linux fedora-linux-24.shared 4.6.7-300.fc24.x86_64) in a Parallels VM on the same machine as the benchmarks above were performed, and the results are as follows:

rsize 131072, wsize 131072: write @ 39.6 mb/s, read @ 111.7 mb/s
rsize 4096, wsize 4096: write @ 30.7 mb/s, read @ 97.5 mb/s

Other than rsize/wsize, The nfs mount flags on Linux used were: rw,bg,hard,resvport,intr,tcp

Therefore there seems to be something that is impacting performance that is specific to the OSX implementation rather than to my machine or environment in general.

**Comment by andrewparks** [comment]
Another FYI for comparison: I've installed gocrypts v1.1 in the same Fedora 24 VM above, and I get the following results:

rsize 131072, wsize 131072: write @ 30.7 mb/s, read @ 125.8 mb/s
rsize 4096, wsize 4096: write @ 20.4 mb/s, read @ 64.5 mb/s

**Comment by vgough** [comment]
On Mac, are you building against a recent version of OSXFuse?

**Comment by andrewparks** [comment]
Thanks for following up - I realised I was running OSXFuse 2.8.3 released Jan 2016. I've now upgraded to the latest 3.5.3. EncFS was installed through Brew - I didn't compile it myself.

Here are the new benchmarks with the latest OSXFuse:

rsize 4096, wsize 4096: write @ 40.2 mb/s, read @ 48.1 mb/s
rsize 8192, wsize 8192: write @ 43.9 mb/s, read @ 64.8 mb/s
rsize 16384, wsize 16384: write @ 53.4 mb/s, read @ 75.3 mb/s
rsize 32768, wsize 32768: write @ 41.3 mb/s, read @ 84.1 mb/s
rsize 131072, wsize 131072: write @ 16.9 mb/s, read @ 93.1 mb/s

So thank you very much, prompting me to upgrade OSXFuse has resulted in a substantial gain - the sweet spot seems to be an rsize/wsize of about 16k
